File: src/presentation/api/contracts/requests/signal_query.py
from dataclasses import dataclass
from datetime import datetime

from src.app.dto import Coin


@dataclass(frozen=True)
class SignalQuery:
    coin: Coin
    start_date: datetime
    end_date: datetime
    signal_names: tuple[str, ...]
    num_signals: int

    def to_params(self) -> list[tuple[str, str | int]]:
        params: list[tuple[str, str | int]] = [
            ("coin", self.coin),
            ("start_date", self.start_date.isoformat()),
            ("end_date", self.end_date.isoformat()),
            ("numSignals", self.num_signals),
        ]
        params.extend(("signalName", name) for name in self.signal_names)
        return params

File: api-dotnet/CryptoTracker.Api/Contracts/Requests/SignalQuery.cs
using Microsoft.AspNetCore.Mvc;

namespace CryptoTracker.Api.Contracts.Requests;

public class SignalQuery
{
    [FromQuery(Name = "coin")]
    public string? Coin { get; init; }

    [FromQuery(Name = "start_date")]
    public DateTimeOffset? StartDate { get; init; }

    [FromQuery(Name = "end_date")]
    public DateTimeOffset? EndDate { get; init; }

    [FromQuery(Name = "signalName")]
    public List<string>? SignalName { get; init; }

    [FromQuery(Name = "numSignals")]
    public int? NumSignals { get; init; }
}
